// cucumber/lib/runtime/step_runner.js
'use strict';Object.defineProperty(exports, "__esModule", { value: true });var _regenerator = require('babel-runtime/regenerator');var _regenerator2 = _interopRequireDefault(_regenerator);var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);var run = function () {var _ref = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(







  function _callee(_ref2) {var attachmentManager = _ref2.attachmentManager,defaultTimeout = _ref2.defaultTimeout,scenarioResult = _ref2.scenarioResult,step = _ref2.step,stepDefinition = _ref2.stepDefinition,transformLookup = _ref2.transformLookup,world = _ref2.world;var parameters, timeoutInMilliseconds, error, result, validCodeLengths, data, attachments, stepResultData;return _regenerator2.default.wrap(function _callee$(_context) {while (1) {switch (_context.prev = _context.next) {case 0:
            beginTiming();
            parameters = stepDefinition.getInvocationParameters({ scenarioResult: scenarioResult, step: step, transformLookup: transformLookup });
            timeoutInMilliseconds = stepDefinition.options.timeout || defaultTimeout;

            error = void 0, result = void 0;
            validCodeLengths = stepDefinition.getValidCodeLengths(parameters);if (!
            _lodash2.default.includes(validCodeLengths, stepDefinition.code.length)) {_context.next = 13;break;}_context.next = 8;return (
              _user_code_runner2.default.run({
                argsArray: parameters,
                fn: stepDefinition.code,
                thisArg: world,
                timeoutInMilliseconds: timeoutInMilliseconds }));case 8:data = _context.sent;

            error = data.error;
            result = data.result;_context.next = 14;break;case 13:

            error = stepDefinition.getInvalidCodeLengthMessage(parameters);case 14:


            attachments = attachmentManager.getAll();
            attachmentManager.reset();

            stepResultData = {
              attachments: attachments,
              duration: endTiming(),
              step: step,
              stepDefinition: stepDefinition };


            if (result === 'pending') {
              stepResultData.status = _status2.default.PENDING;
            } else if (error) {
              stepResultData.failureException = error;
              stepResultData.status = _status2.default.FAILED;
            } else {
              stepResultData.status = _status2.default.PASSED;
            }return _context.abrupt('return',

            new _step_result2.default(stepResultData));case 19:case 'end':return _context.stop();}}}, _callee, this);}));return function run(_x) {return _ref.apply(this, arguments);};}();var _lodash = require('lodash');var _lodash2 = _interopRequireDefault(_lodash);var _status = require('../status');var _status2 = _interopRequireDefault(_status);var _step_result = require('../models/step_result');var _step_result2 = _interopRequireDefault(_step_result);var _time = require('../time');var _time2 = _interopRequireDefault(_time);var _user_code_runner = require('../user_code_runner');var _user_code_runner2 = _interopRequireDefault(_user_code_runner);function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}var beginTiming = _time2.default.beginTiming,endTiming = _time2.default.endTiming;exports.default =


{ run: run };
